HW1/P4/P4.py

```diff
@@ -0,0 +1,32 @@
+from P4_bfs import *
+import findspark
+findspark.init()
+import pyspark
+
+def preprocess_data( filename, sc ):
+    textFile = sc.textFile( filename )
+    textFile.cache()
+    textFile_pre_process = textFile.map( lambda x: ( x.split( "\"" )[3], x.split( "\"" )[1] ) ).partitionBy(num_of_partitions)
+    issue_char_mapping = textFile_pre_process.mapValues( lambda y: [y] ).reduceByKey( lambda a, b: a + b, numPartitions = num_of_partitions )
+    assert copartitioned( textFile_pre_process, issue_char_mapping )
+    char_issue_join = textFile_pre_process.join( issue_char_mapping ).partitionBy(num_of_partitions)
```

Comment:
You could have skipped issue_char_mapping entirely by joining textFile_pre_process with itself. This would also have simplified subsequent processing steps.